packages/dotnet-sdk/src/TeamsFx/Conversation/NotificationMiddleware.cs

```diff
@@ -0,0 +1,130 @@
+
+        public NotificationMiddleware(INotificationTargetStorage storage)
+        {
+            _storage = storage ?? throw new ArgumentNullException(nameof(storage));
+        }
+
+        public async Task OnTurnAsync(ITurnContext turnContext, NextDelegate next, CancellationToken cancellationToken = default)
+        {
+            var activityType = ClassifyActivity(turnContext.Activity);
+            switch (activityType)
+            {
+                case ActivityType.CurrentBotInstalled:
+                case ActivityType.TeamRestored:
+                    {
+                        var reference = turnContext.Activity.GetConversationReference();
+                        await _storage.Write(reference.GetKey(), reference, cancellationToken).ConfigureAwait(false);
+                        break;
+                    }
+                case ActivityType.CurrentBotMessaged:
+                    {
+                        await TryAddMessagedReference(turnContext, cancellationToken).ConfigureAwait(false);
+                        break;
+                    }
+                case ActivityType.CurrentBotUninstalled:
+                case ActivityType.TeamDeleted:
+                    {
+                        var reference = turnContext.Activity.GetConversationReference();
+                        await _storage.Delete(reference.GetKey(), cancellationToken).ConfigureAwait(false);
+                        break;
+                    }
+                default:
+                    {
+                        break;
+                    }
+            }
+
+            await next(cancellationToken).ConfigureAwait(false);
+        }
+
+        private ActivityType ClassifyActivity(Activity activity)
+        {
+            var activityType = activity?.Type;
+            if ("installationUpdate".Equals(activityType, StringComparison.OrdinalIgnoreCase))
+            {
+                var action = activity.Action;
+                if ("add".Equals(action, StringComparison.OrdinalIgnoreCase))
+                {
+                    return ActivityType.CurrentBotInstalled;
+                }
+                else
+                {
+                    return ActivityType.CurrentBotUninstalled;
+                }
+            }
+            else if ("conversationUpdate".Equals(activityType, StringComparison.OrdinalIgnoreCase))
+            {
+                var eventType = activity.GetChannelData<TeamsChannelData>()?.EventType;
+                if ("teamDeleted".Equals(eventType, StringComparison.OrdinalIgnoreCase))
+                {
+                    return ActivityType.TeamDeleted;
+                }
+                else if ("teamRestored".Equals(eventType, StringComparison.OrdinalIgnoreCase))
+                {
+                    return ActivityType.TeamRestored;
+                }
+            }
+            else if ("message".Equals(activityType, StringComparison.OrdinalIgnoreCase))
```

Comment:
You can use the pre-defined type from bot-builder instead of hard-coded string: e.g.: `ActivityTypes.Message`, `ActivityTypes.ConversationUpdate`

Reply:
Good catch. Updated.